### ISSUE TITLE
Securing routes with Firebase Auth access token

### ISSUE THREAD

**Issue by masus04** [tiangolo/fastapi#4768]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

firebase_auth = ...?

@app.get("/")
async def root(token: str = Depends(firebase_auth)):
    return {"message": "Hello World"}
```

### Prior work & resources
From what I read in the fastAPI docs and tutorials, security is implemented using dependency injections and there are some examples on how to do this where the fastAPI server is both the resource as well as the authentication server.

In Issue #12 I found a mention of the `OAuthFlowImplicit` class, which seems about right for my use case. However I do not have enough background knowledge with the involved libraries to solve the issue on my own, so here is my question:

### My task
I have a frontend that authenticates with firebase auth and receives an bearer token. This token is then sent to the backend in which for certain routes I want to check for both authentication and authorization.

So first I need to extract the token from the header and check if it belongs to a valid user, which can be done using the firebase_admin library:
```
firebase_auth.verify_id_token(access_token)
```

Then I need to get the custom claims and search compare them with a list of authorized claims:
```
firebase_admin.auth.UserRecord.custom_claims.any(...)
```

How do I create appropriate dependency injection functions in order to perform these checks?

### Operating System

Linux

### FastAPI Version

0.75.1

### Python Version

3.7 - 3.10

**Comment by jgould22** [comment]
I do not really have any experience with firebase.

So you have a firebase token which is a JWT right?

So you are making a request with JWT which will have a header something like 
```Authorization: Bearer <token>```

which is just a bearer token right ?

So maybe something like this 

```python 
from fastapi import FastAPI
from fastapi.security import HTTPBearer

app = FastAPI()

security = HTTPBearer()

async def firebase_authentication(token: str = Depends(security)):
    decoded_token = auth.verify_id_token(id_token)
    uid = decoded_token['uid']
    # Return the token or throw an exception or whatever 
    return uid

async def firebase_authorization(uid: str = Depends(firebase_authentication)):
    firebase_admin.auth.UserRecord.custom_claims.any(...)
    # Do whatever you gotta do here and return whatever might be useful
    return uid
    
@app.get("/")
async def root(token: uid = Depends(firebase_authorization)):
    return {"message": "Hello World"}
```

I have not tested the above but maybe it can help somewhat

**Comment by masus04** [comment]
> I do not really have any experience with firebase.
> 
> So you have a firebase token which is a JWT right?
> 
> So you are making a request with JWT which will have a header something like `Authorization: Bearer <token>`
> 
> which is just a bearer token right ?
> 
> So maybe something like this
> 
> ```python
> from fastapi import FastAPI
> from fastapi.security import HTTPBearer
> 
> app = FastAPI()
> 
> security = HTTPBearer()
> 
> async def firebase_authentication(token: str = Depends(security)):
>     decoded_token = firebase_admin.verify_id_token(id_token)
>     uid = decoded_token['uid']
>     # Return the token or throw an exception or whatever 
>     return uid
> 
> async def firebase_authorization(uid: str = Depends(firebase_authentication)):
>     firebase_admin.auth.UserRecord.custom_claims.any(...)
>     # Do whatever you gotta do here and return whatever might be useful
>     return uid
>     
> @app.get("/")
> async def root(uid: str = Depends(firebase_authorization)):
>     return {"message": "Hello World"}
> ```
> 
> I have not tested the above but maybe it can help somewhat

This looks promising, thank you for your answer!

I added some firebase implementation to it and came up with the following:
```
security = HTTPBearer()

api = FastAPI()
security = HTTPBearer()

firebase_client = FirebaseClient(
    firebase_admin_credentials_url=firebase_test_admin_credentials_url
    # ...
)

user_roles = [test_role]

async def firebase_authentication(token: HTTPAuthorizationCredentials = Depends(security)) -> dict:
    user = firebase_client.verify_token(token.credentials)
    return user

async def firebase_authorization(user: dict = Depends(firebase_authentication)):
    roles = firebase_client.get_user_roles(user)

    for role in roles:
        if role in user_roles:
            return user

    raise HTTPException(detail="User does not have the required roles", status_code=HTTPStatus.FORBIDDEN)

@api.get("/")
async def root(uid: str = Depends(firebase_authorization)):
    return {"message": "Successfully authenticated & authorized!"}
```

Using a http client like postman, this seems to work perfectly, however when trying it out using the OpenApi docs I noticed that a 403 error is correctly returned, since no token is passed in the header.
It seems like the OpenApi docs does not know about the required header. Is there a way to configure HTTPBearer in such a way that this information is inferred?

**Comment by jgould22** [comment]
You probably need to add your jwt with the Authorize button in the top right of the open api interface. 
![image](https://user-images.githubusercontent.com/4335847/162464561-1c54551d-9a77-444c-8026-d6476fda2d70.png)

**Comment by masus04** [comment]
Seems reasonable, however my UI does not have such a button :)

Any idea how I configure FastApi to let me authorize?

**Comment by jgould22** [comment]
Oh? That is odd. I did not do anything special to get the button to appear, has your browser cached the openapi.json?

When I run the below sample I get the button with no further configuration 
```python 
from fastapi import FastAPI, Depends
from fastapi.security import HTTPBearer

app = FastAPI()

security = HTTPBearer()

async def firebase_authentication(token: str = Depends(security)):

    return "test"

async def firebase_authorization(uid: str = Depends(firebase_authentication)):
    return uid
    
@app.get("/")
async def root(uid: str = Depends(firebase_authorization)):
    return {f"message": "Hello World - {uid}"}

**Comment by masus04** [comment]
I'm sorry, I was running another version that did not have the `HTTPBearer`, hence no button.

Now it's working great, thank you very much :+1:

**Comment by masus04** [comment]
One more question: Is there any way to computationally turn Bearer auth on/off?

I tried to do it inside the authentication methods, but if no token is passed, a `403 error` is returned before these are even executed.

**Comment by jgould22** [comment]
The code that runs the HTTPBearer is here https://github.com/tiangolo/fastapi/blob/master/fastapi/security/http.py#L100

You can see it will throw a 403 if you do not include the token.

As such using the HTTPBearer probably would not allow you to computationally turn it on and off.

You could take that class and subclass it or make your own that does have some ability to enable or disable it.

**Comment by masus04** [comment]
> 

Exactly. The scenario is that I am using Bearer Auth with firebase auth and would like to be able to turn it off for most test cases. However, as soon as the HTTPBearer is instantiated, any request will fail if the required header is not present, even before theauth service can ever be called.

**Comment by adriangb** [comment]
Can't you just include the header in your tests?

```python
from fastapi import APIRouter, Depends, FastAPI
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

router = APIRouter()

security = HTTPBearer()


async def validate_firebase_token(token: HTTPAuthorizationCredentials = Depends(security)) -> None:
    ...


@router.get("/", dependencies=[Depends(validate_firebase_token)])
def endpoint() -> None:
    ...



def create_app() -> FastAPI:
    app = FastAPI()
    app.include_router(router)
    return app


from fastapi.testclient import TestClient


def test_endpoint() -> None:

    test_token = "foo"
    
    async def validate_fake_token(token: HTTPAuthorizationCredentials = Depends(security)) -> None:
        assert token.credentials == test_token

    app = create_app()
    app.dependency_overrides[validate_firebase_token] = validate_fake_token

    client = TestClient(app)

    resp = client.get("/", headers={"Authorization": f"Bearer {test_token}"})
    assert resp.status_code == 200


test_endpoint()
```

**Comment by jgould22** [comment]
Additionally if you are using Pytest you can use its Monkey patch fixture to override any function on a per test basis which can be very useful.

**Comment by masus04** [comment]
@adriangb I assume that's the most straight forward way to do it.

Pass a placeholder token so that the HTTBearer does not throw an Exception, then add a parameter `auth_enabled` to the validate_firebase_token method, depending on which we either call the auth service or don't do anything.

Thx :+1:
